lib/Ethereum/RPC/Contract.pm

```diff
@@ -106,149 +132,164 @@ sub invoke {
-
 }
 
 =head2 _prepare_transaction
 
 Join the data and parameters and return a prepared transaction to be called as send, call or deploy.
 
-Parameters:
-    $compiled_data  ( Required - the hashed function string name with parameters or the compiled contract bytecode )
-    params          ( Required - the parameters args given by the method call )
+=over 4
+
+=item compiled_data => string (required, function signature or the contract bytecode)
+
+=item params => array (required)
 
-Return:
-    Future object
-        on_done: Ethereum::Contract::ContractTransaction
-        on_fail: string message
+=back
+
+L<Future> object
+on_done: L<Ethereum::Contract::ContractTransaction>
+on_fail: error string
 
 =cut
 
 sub _prepare_transaction {
     my ($self, $compiled_data, $params) = @_;
 
-    my $data = join("", $compiled_data, map { $self->get_hex_param($_) } @{$params});
+    my $hex_params = $self->get_hex_param($params);
+
+    my $data = $compiled_data . $hex_params;
 
     return Ethereum::RPC::Contract::ContractTransaction->new(
-        contract_address=> $self->contract_address,
-        rpc_client      => $self->rpc_client,
-        data            => $self->append_prefix($data),
-        from            => $self->from,
-        gas             => $self->gas,
-        gas_price       => $self->gas_price,
+        contract_address => $self->contract_address,
+        rpc_client       => $self->rpc_client,
+        data             => $self->append_prefix($data),
+        from             => $self->from,
+        gas              => $self->gas,
+        gas_price        => $self->gas_price,
     );
 
 }
 
 =head2 get_hex_param
 
-Convert the given value to hexadecimal format
+Convert parameter list to the ABI format:
+https://solidity.readthedocs.io/en/develop/abi-spec.html#function-selector-and-argument-encoding
+
+=over 4
 
-Parameters:
-    function_id (Required - arg to be converted to hexadecimal)
+=item params => array (required)
 
-Return:
-    Hexadecimal string
+=back
+
+Returns a string containing the ABI format to be send to the contract.
 
 =cut
 
 sub get_hex_param {
-    my ($self, $param) = @_;
-
-    my $new_param;
-    # Is hexadecimal string
-    if( $param =~ /^0x[0-9A-F]+$/i ) {
-        $new_param = sprintf( "%064s", substr($param, 2) );
-    # Is integer
-    } elsif ( $param =~ /^[+-]?[0-9?e+]+$/ ) {
-        $new_param = sprintf( "%064s", substr(Math::BigInt->new($param)->as_hex, 2) );
-    # Is string
-    } else {
-        $new_param = sprintf( "%064s", unpack("H*", $param) );
+    my ($self, $params) = @_;
+
+    my @offset_indices;
+    my @static;
+    my @dynamic;
+
+    #TODO:
+    # - Arrays
+    # - Bytes
+    for my $param (@$params) {
+        if ($param =~ /^0x[0-9A-F]+$/i) {
+            push(@static, sprintf("%064s", substr($param, 2)));
+        } elsif (looks_like_number($param)) {
+            push(@static, sprintf("%064s", substr(Math::BigInt->new($param)->as_hex, 2)));
```

Comment:
Probably something to revisit later: as of Perl 5.26 and updated cpan, we now have Math::BigInt 1.999811 which has a [`to_hex()`](https://metacpan.org/pod/Math::BigInt#to_hex()) method, which can greatly simplify things as I notice we still do some `0x` fudging elsewhere.